src/forder.c

```diff
@@ -440,11 +440,62 @@ SEXP forder(SEXP DT, SEXP by, SEXP retGrpArg, SEXP sortGroupsArg, SEXP ascArg, S
   nrow = length(VECTOR_ELT(DT,0));
+  int n_cplx = 0;
   for (int i=0; i<LENGTH(by); i++) {
-    if (INTEGER(by)[i] < 1 || INTEGER(by)[i] > length(DT))
-      error("'by' value %d out of range [1,%d]", INTEGER(by)[i], length(DT));
-    if ( nrow != length(VECTOR_ELT(DT, INTEGER(by)[i]-1)) )
+    int by_i = INTEGER(by)[i];
+    if (by_i < 1 || by_i > length(DT))
+      error("'by' value %d out of range [1,%d]", by_i, length(DT));
+    if ( nrow != length(VECTOR_ELT(DT, by_i-1)) )
       error("Column %d is length %d which differs from length of column 1 (%d)\n", INTEGER(by)[i], length(VECTOR_ELT(DT, INTEGER(by)[i]-1)), nrow);
+    if (TYPEOF(VECTOR_ELT(DT, by_i-1)) == CPLXSXP) n_cplx++;
+  }
+  if (n_cplx) {
+    // we don't expect users to need complex sorting extensively
+    //   or on massive data sets, so we take the approach of
+    //  splitting a complex vector into its real & imaginary parts
+    //  and using the regular forderv machinery to sort; a baremetal
+    //  implementation would at root do the same, but the approach
+    //  here is a bit more slapdash with respect to memory efficiency
+    //  (seen clearly here at C from the 3+2*n_cplx PROTECT() calls)
+    int n_out = length(by) + n_cplx;
+    SEXP new_dt = PROTECT(allocVector(VECSXP, n_out)); n_protect++;
+    SEXP new_asc = PROTECT(allocVector(INTSXP, n_out)); n_protect++;
+    // will be simply 1:n_out
+    SEXP new_by = PROTECT(allocVector(INTSXP, n_out)); n_protect++;
+    int j = 0;
+    for (int i=0; i<length(by); i++) {
+      int by_idx = INTEGER(by)[i]-1;
+      if (TYPEOF(VECTOR_ELT(DT, by_idx)) == CPLXSXP) {
+        // I don't see any shorthand way of splitting of the real&imaginary components,
+        //   i.e., a shorthand way of doing Re(z), Im(z). That includes searching all of
+        //   the r-source code & all of the r-devel archives. So just reproduce Re(), Im()
+        //   as done in do_cmathfuns in complex.c
+        SEXP realPart = PROTECT(allocVector(REALSXP, nrow)); n_protect++;
+        SEXP cplxPart = PROTECT(allocVector(REALSXP, nrow)); n_protect++;
+        double *pre = REAL(realPart);
+        double *pim = REAL(cplxPart);
+        Rcomplex *pz = COMPLEX(VECTOR_ELT(DT, by_idx));
+        for (int i = 0; i < nrow; i++) {
+          pre[i] = pz[i].r;
+          pim[i] = pz[i].i;
+        }
+        SET_VECTOR_ELT(new_dt, j, realPart);
+        SET_VECTOR_ELT(new_dt, j+1, cplxPart);
+        INTEGER(new_asc)[j] = INTEGER(ascArg)[i];
+        INTEGER(new_asc)[j+1] = INTEGER(ascArg)[i];
+        INTEGER(new_by)[j] = j+1;
+        INTEGER(new_by)[j+1] = j+2;
+        j += 2;
+      } else {
+        SET_VECTOR_ELT(new_dt, j, VECTOR_ELT(DT, by_idx));
+        INTEGER(new_asc)[j] = INTEGER(ascArg)[i];
+        INTEGER(new_by)[j] = j+1;
+        j += 1;
+      }
+    }
+    DT = new_dt;
+    ascArg = new_asc;
+    by = new_by;
   }
```

Comment:
I'm a bit nervous about this branch.  I looks constrained to only when n_cplx>0 (which is good that it's isolated), and I know _what_ it is doing (splitting complex into real and imaginary columns) but I don't see _how_ it's doing it. I'm going to need more time to review it and make sure it doesn't impact anything unexpected.

Reply:
How about dealing with complex columns further down in this switch.  (GitHub wouldn't let me add a comment on that line as it's outside the diff perhaps? so I pasted an image)
![image](https://user-images.githubusercontent.com/2779998/61498003-5ba0a300-a976-11e9-8062-fcf7831fe164.png)

Reply:
when i first took a look i decided against that (I think my initial reaction was `range` doesn't make sense for complex, but of course it does because they're well-ordered using the lexicographic rule). will take another look.

The `n_cplx` branch is a bit messy, and took me several iterations to get there as I kept failing a few more tests each time. I think it can be cleaned up a little by snazzier use of the `j` index.

Idea is: receive `by` which is say length `n`, `m` of which are complex columns.

Instead of sorting using the input list `DT`, do so using a new list (constructed as `new_dt`).

If there are no complex columns, `new_dt[[1]]` is `DT[[ by[1] ]]`, `new_dt[[2]]` is `DT[[ by[2] ]]`, and so on.

If there _are_ complex columns, say the first is at the third index of `by`, then `new_dt[[3]]` is `Re( DT[[ by[3] ]] )` and `new_dt[[4]]` is `Im( DT[[ by[3] ]] )`. Then continuing on as usual. That "bump" of the index by a complex column is what made things complicated.

Reply:
I wasn't thinking of doing a single range for `complex`.  The thought was to trick the loop into doing a complex column twice. The first time doing range for the real part only. The 2nd time calling range for the imaginary part only.  There might need to be new range_complex_r and range_complex_i perhaps.

Reply:
The other option is to postpone this PR to a later date. Elio's response above confirms that setkey-on-complex and grouping-on-complex isn't high priority. It was `complex` value columns that was the main thing;  which you did and merged.

Reply:
Sure. The main thing done in this PR is to close #1703 (grouping/keying/others mentioned here come almost-for-free after that). And this branch is the workhorse for that. Could `wontfix`&close that issue I guess.

Reply:
> I wasn't thinking of doing a single range for complex. The thought was to trick the loop into doing a complex column twice. The first time doing range for the real part only. The 2nd time calling range for the imaginary part only. There might need to be new range_complex_r and range_complex_i perhaps.

Oh, I think I see now. The idea is, instead of bracketing `[re_min, re_max]` and `[im_min, im_max]` separately, to just bracket both components with `[min(re_min, im_min), max(re_max, im_max)]`

Reply:
Not sure what you mean by bracketing.  But I meant treating the complex column just as if it were two separate columns of double.

Reply:
It would nice for completeness to get this merged and closed #1703, yes. But I'm also an eye that we've both been working on complex number support for 3 days now (elapsed) and continuing. And Elio said going this far (sorting complex) isn't needed. There isn't any demand for this from users as far as I can see.

Reply:
I think the biggest impediment to that is I don't see a way to get a `double *` object pointing to the real/imaginary _vectors_ (like `REAL(COMPLEX(x))` or `COMPLEX(x).r`, hope that makes sense), which is what all of the functions for `REAL` columns take as input.

Reply:
> It would nice for completeness to get this merged and closed #1703, yes. But I'm also an eye that we've both been working on complex number support for 3 days now (elapsed) and continuing. And Elio said going this far (sorting complex) isn't needed. There isn't any demand for this from users as far as I can see.

True... I'm just loath to close the PR given that it's passing tests. Also understand the tradeoff that is the maintenance burden.

Reply:
> I think the biggest impediment to that is I don't see a way to get a double * object pointing to the real/imaginary vectors (like REAL(COMPLEX(x)) or COMPLEX(x).r, hope that makes sense), which is what all of the functions for REAL columns take as input.

Yes makes sense. That's why I wrote above: "There might need to be new range_complex_r and range_complex_i perhaps."   But I didn't look at it closely.

> maintenance burden

It's not a burden in the sense that more code is just always generally bad.  It's that this particular change could cause hard to trace bugs because of the type of change it is to such a core function.  It's risk.  And it's time in reviewing it.

> it's passing tests

But this shows the tests aren't sufficient: the cast isn't correct but it's still passing tests.

Reply:
> But this shows the tests aren't sufficient: the cast isn't correct but it's still passing tests.

Very true, I meant more that it's passing _existing_ tests (so "no" existing functionality is being affected).

Would adding a lot more tests to the suite for the new code be a good direction?

Reply:
A good direction would be to focus on the most requested issues: #3189